NEW: Add CS Protocol /r0/rooms/{roomId}/context/{eventId}

kevinrademan opened the pull request:
Adds support for a missing api method from the CS protocol 
https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-context-eventid

kevinrademan commented:
Side note, in order for the filter to work you need to to pass it along as below. While this is consistent with other api endpoints, it might not be the most intuitive code. Perhaps filter input objects should be parsed and automatically converted depending on the input type?

```
filter = { types: ['m.room.message'] }
api.get_room_event_context("roomid","eventid", filter: filter.to_json)
```

ananace commented on lib/matrix_sdk/protocols/cs.rb:
```diff
@@ -767,6 +767,28 @@ def get_room_state_all(room_id, **params)
+  # Retrieves number of events that happened just before and after the specified event
+  #
+  # @param room_id [MXID,String] The room to get events from.
+  # @param event_id [String] The event to get context around.
```
Nit: Should probably be `[MXID,String]` here too, event IDs are also a kind of MXID

ananace commented on lib/matrix_sdk/protocols/cs.rb:
```diff
@@ -767,6 +767,28 @@ def get_room_state_all(room_id, **params)
+  #
+  # @param room_id [MXID,String] The room to get events from.
+  # @param event_id [String] The event to get context around.
+  # @param limit [Integer] (10) The limit of messages to retrieve
```
This should be documented as an `@option` as it's part of the `params` hash just like `:filter`.

ananace reviewed (changes requested):
It might be a good idea to add some documentation on how the filter is expected to be passed, perhaps with an `@example` of your comment.

On the note of said comment; I've been trying to keep the `Api` object itself doing as few input modifications as possible, but I'll have a think on maybe defaulting to built-in `#to_json` calls iff: the input is a hash, it's passed as a query parameter, and it's expected to contain JSON

kevinrademan commented:
I've fixed up those 2 documentation types and added a basic example. That api endpoint returns a pretty big response, so I didnt think it would be appropriate to include that in the example itself. The matrix documentation doesn however have a good example of the response payload

ananace commented on lib/matrix_sdk/protocols/cs.rb:
```diff
@@ -767,6 +767,30 @@ def get_room_state_all(room_id, **params)
+  # @see https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-context-eventid
+  #      The Matrix Spec, for more information about the call and response
+  # @example Find event context with filter and limit specified
+  #   api.get_room_event_context('##room_id##', '##event_id##', filter: { types: ['m.room.message'] }.to_json, limit: 20)
```
I'd suggest using things like `#room:example.com` and `$event_id:example.com` or `$exampleMatrixEventId` as the examples, to use the same grammar as what the input is expected to be.

ananace commented:
Yeah, some of the Matrix responses are quite immense, no need to include all - or even most - of them in the examples. I tend to stick to only providing the main data point that the example is trying to show, with many a key or two of context.

A plain example that simply shows off what to pass to the method is more than good enough.

ananace commented on lib/matrix_sdk/protocols/cs.rb:
```diff
@@ -767,6 +767,30 @@ def get_room_state_all(room_id, **params)
+  #
+  # @param room_id [MXID,String] The room to get events from.
+  # @param event_id [MXID,String] The event to get context around.
+  # @option limit [Integer] (10) The limit of messages to retrieve
```
This should probably be `@option params [Integer] :limit (10)`, the first part - the variable - is the source of the option, then you have the type, then the name of the option `:limit`, and then the default value

Ref: https://rubydoc.info/gems/yard/file/docs/Tags.md#option

kevinrademan replied:
ahh yeah thats my bad. I didnt fixup the definition after changing this from an @param to an @option

ananace reviewed (changes requested):
And one final piece and everything should be just fine.
